src/symbol/placement.js

```diff
@@ -0,0 +1,414 @@
+// @flow
+
+const CollisionIndex = require('./collision_index');
+const EXTENT = require('../data/extent');
+const symbolSize = require('./symbol_size');
+const projection = require('./projection');
+const symbolLayoutProperties = require('../style/style_layer/symbol_style_layer_properties').layout;
+const assert = require('assert');
+const pixelsToTileUnits = require('../source/pixels_to_tile_units');
+
+import type Transform from '../geo/transform';
+import type StyleLayer from '../style/style_layer';
+import type Tile from '../source/tile';
+import type SymbolBucket from '../data/bucket/symbol_bucket';
+import type mat4 from '@mapbox/gl-matrix';
+import type {CollisionBoxArray, CollisionVertexArray} from '../data/array_types';
+
+class OpacityState {
+    opacity: number;
+    placed: boolean;
+    constructor(prevState: ?OpacityState, increment: number, placed: boolean, offscreen: ?boolean) {
+        if (prevState) {
+            this.opacity = Math.max(0, Math.min(1, prevState.opacity + (prevState.placed ? increment : -increment)));
+        } else {
+            this.opacity = (offscreen && placed) ? 1 : 0;
+        }
+        this.placed = placed;
+    }
+    isHidden() {
+        return this.opacity === 0 && !this.placed;
+    }
+}
+
+class JointOpacityState {
+    text: OpacityState;
+    icon: OpacityState;
+    constructor(prevState: ?JointOpacityState, increment: number, placedText: boolean, placedIcon: boolean, offscreen: ?boolean) {
+        this.text = new OpacityState(prevState ? prevState.text : null, increment, placedText, offscreen);
+        this.icon = new OpacityState(prevState ? prevState.icon : null, increment, placedIcon, offscreen);
+    }
+    isHidden() {
+        return this.text.isHidden() && this.icon.isHidden();
+    }
+}
+
+class JointPlacement {
+    text: boolean;
+    icon: boolean;
+    // offscreen = outside viewport, but within CollisionIndex::viewportPadding px of the edge
```

Comment:
By the time it gets here "offscreen" now means something like "skip fade animation" since it's the combination of "offscreen" and "justReloaded". We should come up with a new name to account for that once the two concepts are combined.